// @sendbird/uikit-react/cjs/hooks/useConnectionState.js
'use strict';

var React = require('react');
var SendbirdChat = require('@sendbird/chat');
var uuid = require('../chunks/bundle-KfgcZyeV.js');
var useSendbird = require('../chunks/bundle-BMIzPycb.js');
require('../chunks/bundle-DqhRrRu9.js');
require('../chunks/bundle-BotH0J8Z.js');
require('../chunks/bundle-DBIZFnYM.js');
require('@sendbird/chat/groupChannel');
require('@sendbird/chat/openChannel');

var useConnectionState = function () {
    var stores = useSendbird.useSendbird().state.stores;
    var sdkStore = stores.sdkStore;
    var sdk = sdkStore.sdk;
    var _a = React.useState(sdk.connectionState), connectionState = _a[0], setConnectionState = _a[1];
    React.useEffect(function () {
        var _a;
        var handlerId = uuid.uuidv4();
        (_a = sdk === null || sdk === void 0 ? void 0 : sdk.addConnectionHandler) === null || _a === void 0 ? void 0 : _a.call(sdk, handlerId, new SendbirdChat.ConnectionHandler({
            onConnected: function () { return setConnectionState(SendbirdChat.ConnectionState.OPEN); },
            onDisconnected: function () { return setConnectionState(SendbirdChat.ConnectionState.CLOSED); },
            onReconnectStarted: function () { return setConnectionState(SendbirdChat.ConnectionState.CONNECTING); },
            onReconnectSucceeded: function () { return setConnectionState(SendbirdChat.ConnectionState.OPEN); },
            onReconnectFailed: function () { return setConnectionState(SendbirdChat.ConnectionState.CLOSED); },
        }));
        return function () {
            var _a;
            (_a = sdk === null || sdk === void 0 ? void 0 : sdk.removeConnectionHandler) === null || _a === void 0 ? void 0 : _a.call(sdk, handlerId);
        };
    }, [sdk]);
    return connectionState;
};

exports.useConnectionState = useConnectionState;
